Sources/GRPCCore/RuntimeError.swift

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// An error thrown at runtime.
+///
+/// In contrast to ``RPCError``, the ``RuntimeError`` represents errors which happen at a scope
+/// wider than an individual RPC. For example, passing invalid configuration values.
+public struct RuntimeError: Error, Hashable, @unchecked Sendable {
```

Comment:
It _might_ make sense to replace the `ClientError` and `ServerError` with this. They all have the same structure but differ only in error codes. WDYT @gjcairo?

Reply:
Hmm my first guess would be it makes sense, but I do wonder if it would be useful from a user's perspective to be able to differentiate between client and server errors beyond the error codes. I'm thinking something like:
```swift
func wrapAndRecordMetrics(someClosure: () throws -> Void) {
  do {
    try someClosure()
  } catch is ServerError {
    // increase server error metric counter
  } catch is ClientError {
    // increase client error metric counter
  } catch {
    throw error
  }
}
```

But I'm not sure how realistic something like this would be, and I can't come up with any other obvious scenarios, so maybe it would make sense to replace them.

Reply:
Forgot to reply to this one earlier: I don't know if that's useful because we now a third `RuntimeError` which can be thrown by either client or server so this sort of metric recording fails to be useful anyway.